packages/consensus/src/types/utreexo.cairo

```diff
@@ -72,8 +77,87 @@ pub trait UtreexoAccumulator {
     fn delete_batch(ref self: UtreexoState, proof: @UtreexoBatchProof);
 }
 
-#[derive(Drop, Copy, PartialEq)]
-pub enum UtreexoError {}
+pub impl UtreexoStateImpl of UtreexoAccumulator {
+    /// Adds single output to the accumulator.
+    /// The order *is important*: adding A,B and B,A would result in different states.
+    ///
+    /// Note that this call also pushes old UTXOs "to the left", to a larger subtree.
+    /// This mechanism ensures that short-lived outputs have small inclusion proofs.
+    fn add(ref self: UtreexoState, output: OutPoint) {}
+
+    /// Verifies inclusion proof for a single output.
+    fn verify(
+        self: @UtreexoState, output: @OutPoint, proof: @UtreexoProof
+    ) -> Result<(), UtreexoError> {
+        let txid_u256: u256 = (*output.txid).into();
+        let proof_root = verify_helper(
+            *proof.proof, *proof.leaf_index, txid_u256.try_into().expect('value too large')
+        );
+
+        // Get the expected root
+        let root_index = (*proof.proof).len();
+        let expected_root = self.roots[root_index];
+
+        match expected_root {
+            Option::Some(root) => {
+                if (*root).into() == proof_root {
+                    return Result::Ok(());
+                };
+                Result::Err(UtreexoError::ProofVerificationFailed)
+            },
+            _ => Result::Err(UtreexoError::ProofVerificationFailed)
+        }
+    }
+
+
+    /// Removes single output from the accumlator (order is important).
+    ///
+    /// Note that once verified, the output itself is not required for deletion,
+    /// the leaf index plus inclusion proof is enough.
+    fn delete(ref self: UtreexoState, proof: @UtreexoProof) {}
+
+    /// Verifies batch proof for multiple outputs (e.g. all outputs in a block).
+    fn verify_batch(
+        self: @UtreexoState, outputs: Span<OutPoint>, proof: @UtreexoBatchProof
+    ) -> Result<(), UtreexoError> {
+        Result::Ok(())
+    }
+
+    /// Removes multiple outputs from the accumulator.
+    fn delete_batch(ref self: UtreexoState, proof: @UtreexoBatchProof) {}
+}
+
+fn verify_helper(proof: Span<felt252>, mut leaf_index: u64, mut curr_node: felt252) -> felt252 {
+    let mut proof = proof.clone();
+    let proof_length = proof.len();
+
+    if proof_length == 0 {
+        return curr_node;
+    }
+
+    for _ in 0
+        ..proof_length {
+            let (next_left_index, is_right) = DivRem::div_rem(leaf_index, TWO);
```

Comment:
😄 a constant for 2 is too much:)

Reply:
I can't figure how to pass an argument to `div_rem`, `DivRem::div_rem(leaf_index, 2)` or `DivRem::div_rem(leaf_index, 2.into()` not work, and u64_to_nonzero isn't importable.

Reply:
Try 2_u32?

Reply:
`Unexpected argument type. Expected: "core::zeroable::NonZero::<core::integer::u64>", found: "core::integer::u64".`, something like this

Reply:
Tested on your branch and `DivRem::div_rem(leaf_index, 2)` seems to work fine